pkg/sql/create_type.go

```diff
@@ -99,6 +101,91 @@ func getCreateTypeParams(
 	return typeKey, id, nil
 }
 
+// When a type is created in Postgres, Postgres will implicitly create an array
+// type of that user defined type. This array type tracks changes to the
+// original type, and is dropped when the original type is dropped.
+// createArrayType creates the implicit array type for the input TypeDescriptor
+// and returns the ID of the created type.
+func (p *planner) createArrayType(
+	params runParams,
+	n *tree.CreateType,
+	typ *tree.TypeName,
+	typDesc *sqlbase.MutableTypeDescriptor,
+	db *DatabaseDescriptor,
+) (sqlbase.ID, error) {
+	// Postgres starts off trying to create the type as _<typename>. It then
+	// continues adding "_" to the front of the name until it doesn't find
+	// a collision.
+	schemaID := sqlbase.ID(keys.PublicSchemaID)
+	arrayTypeName := "_" + typ.Type()
+	var arrayTypeKey sqlbase.DescriptorKey
+	for {
+		// See if there is a collision with the current name.
+		exists, _, err := sqlbase.LookupObjectID(
+			params.ctx,
+			params.p.txn,
+			params.ExecCfg().Codec,
+			db.ID,
+			schemaID,
+			arrayTypeName,
+		)
+		if err != nil {
+			return 0, err
+		}
+		// If we found an empty spot, then create the namespace key for this entry.
+		if !exists {
+			arrayTypeKey = sqlbase.MakePublicTableNameKey(
+				params.ctx,
+				params.ExecCfg().Settings,
+				db.ID,
+				arrayTypeName,
+			)
+			break
+		}
+		// Otherwise, append another "_" to the front of the name.
+		arrayTypeName = "_" + arrayTypeName
+	}
+
+	// Generate the stable ID for the array type.
+	id, err := catalogkv.GenerateUniqueDescID(params.ctx, params.ExecCfg().DB, params.ExecCfg().Codec)
+	if err != nil {
+		return 0, err
+	}
+
+	// Create the element type for the array. Note that it must know about the
+	// ID of the array type in order for the array type to correctly created.
+	var elemTyp *types.T
+	switch t := typDesc.Kind; t {
+	case sqlbase.TypeDescriptor_ENUM:
+		elemTyp = types.MakeEnum(uint32(typDesc.ID), uint32(id))
+	default:
+		return 0, errors.AssertionFailedf("cannot make array type for kind %s", t.String())
+	}
+
+	// Construct the descriptor for the array type.
+	arrayTypDesc := sqlbase.NewMutableCreatedTypeDescriptor(sqlbase.TypeDescriptor{
```

Comment:
I think we should make a type descriptor constructor. Aren't we making these things all over the place? If we ever add a new field, we'd have to go back and edit all of the places that make a raw struct.

Reply:
done